File: lib/features/remix/application/remix_telemetry.dart
import 'package:flutter/foundation.dart';

import 'package:music_director/features/remix/domain/remix_descriptors.dart';
import 'package:music_director/features/remix/domain/remix_mode.dart';

/// Privacy-safe remix observability — never logs raw title/artist.
abstract final class RemixTelemetry {
  RemixTelemetry._();

  static void emit(Map<String, Object?> fields) {
    final parts = <String>[
      for (final e in fields.entries)
        if (e.value != null) '${e.key}=${e.value}',
    ];
    debugPrint('[remix] ${parts.join(' ')}');
  }

  static Map<String, Object?> activationFields({
    required RemixMode mode,
    required bool nearActivation,
    required String genre,
    required String songGenerationType,
    String title = '',
    String artist = '',
    bool blockInjected = false,
  }) {
    return {
      'event': 'activation',
      'mode': mode.name,
      'near': nearActivation,
      'genre': genre.trim().isEmpty ? 'unset' : genre.trim(),
      'gen': songGenerationType,
      'block_v': remixBlockVersion,
      'fp': (title.trim().length >= 2 && artist.trim().length >= 2)
          ? RemixDescriptors.fingerprint(title, artist)
          : 'none',
      'injected': blockInjected,
    };
  }

  static void activation({
    required RemixMode mode,
    required bool nearActivation,
    required String genre,
    required String songGenerationType,
    String title = '',
    String artist = '',
    bool blockInjected = false,
  }) {
    emit(
      activationFields(
        mode: mode,
        nearActivation: nearActivation,
        genre: genre,
        songGenerationType: songGenerationType,
        title: title,
        artist: artist,
        blockInjected: blockInjected,
      ),
    );
  }

  static Map<String, Object?> postProcessFields({
    required RemixMode mode,
    required String songGenerationType,
    required int strippedLyricLines,
    required int stagingInjected,
    required bool leakHit,
    String title = '',
    String artist = '',
  }) {
    return {
      'event': 'post_process',
      'mode': mode.name,
      'gen': songGenerationType,
      'strip_n': strippedLyricLines,
      'staging_n': stagingInjected,
      'leak': leakHit,
      'block_v': remixBlockVersion,
      'fp': (title.trim().length >= 2 && artist.trim().length >= 2)
          ? RemixDescriptors.fingerprint(title, artist)
          : 'none',
    };
  }

  static void postProcess({
    required RemixMode mode,
    required String songGenerationType,
    required int strippedLyricLines,
    required int stagingInjected,
    required bool leakHit,
    String title = '',
    String artist = '',
  }) {
    emit(
      postProcessFields(
        mode: mode,
        songGenerationType: songGenerationType,
        strippedLyricLines: strippedLyricLines,
        stagingInjected: stagingInjected,
        leakHit: leakHit,
        title: title,
        artist: artist,
      ),
    );
  }
}

File: server/app/remix_telemetry.py
"""Privacy-safe remix observability — never logs raw title/artist."""

from __future__ import annotations

import logging
from typing import Any

from app.remix_engine import REMIX_BLOCK_VERSION, RemixMode, remix_source_fingerprint

logger = logging.getLogger("music_director.remix")


def emit_remix_event(**fields: Any) -> None:
    parts = [f"{k}={v}" for k, v in fields.items() if v is not None]
    logger.info("[remix] %s", " ".join(parts))


def log_remix_activation(
    *,
    mode: RemixMode,
    near_activation: bool,
    genre: str,
    song_generation_type: str,
    title: str = "",
    artist: str = "",
    block_injected: bool = False,
) -> None:
    fp = (
        remix_source_fingerprint(title, artist)
        if len(title.strip()) >= 2 and len(artist.strip()) >= 2
        else "none"
    )
    emit_remix_event(
        event="activation",
        mode=mode.value,
        near=near_activation,
        genre=(genre.strip() or "unset"),
        gen=song_generation_type,
        block_v=REMIX_BLOCK_VERSION,
        fp=fp,
        injected=block_injected,
    )


def log_remix_post_process(
    *,
    mode: RemixMode,
    song_generation_type: str,
    stripped_lyric_lines: int,
    staging_injected: int,
    leak_hit: bool,
    title: str = "",
    artist: str = "",
) -> None:
    fp = (
        remix_source_fingerprint(title, artist)
        if len(title.strip()) >= 2 and len(artist.strip()) >= 2
        else "none"
    )
    emit_remix_event(
        event="post_process",
        mode=mode.value,
        gen=song_generation_type,
        strip_n=stripped_lyric_lines,
        staging_n=staging_injected,
        leak=leak_hit,
        block_v=REMIX_BLOCK_VERSION,
        fp=fp,
    )
